pkg/dyninst/decode/marshal.go

```diff
@@ -0,0 +1,473 @@
+			); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndObject); err != nil {
+			return err
+		}
+
+	case *ir.ArrayType:
+		var err error
+		if err = writeTokens(enc, jsontext.BeginArray); err != nil {
+			return err
+		}
+
+		elementType := v.Element
+		elementSize := int(elementType.GetByteSize())
+		numElements := int(v.Count)
+		for i := range numElements {
+			elementData := data[i*elementSize : (i+1)*elementSize]
+			if err := d.encodeValue(enc,
+				dataItems, currentlyEncoding, v.Element, elementData); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndArray); err != nil {
+			return err
+		}
+		return nil
+	case *ir.GoEmptyInterfaceType:
+		return writeTokens(enc, jsontext.String("unimplemented - empty interface"))
+	case *ir.GoInterfaceType:
+		return writeTokens(enc, jsontext.String("unimplemented - interface type"))
+	case *ir.GoSliceHeaderType:
+		if len(data) < int(v.ByteSize) {
+			return errors.New("passed data not long enough for slice header")
+		}
+		address := binary.NativeEndian.Uint64(data[0:8])
+		length := binary.NativeEndian.Uint64(data[8:16])
+		if length == 0 {
+			if err := writeTokens(enc,
+				jsontext.BeginArray,
+				jsontext.EndArray); err != nil {
+				return err
+			}
+			return nil
+		}
+		elementType := v.Data.ID
+		elementSize := int(v.Data.Element.GetByteSize())
+		sliceDataItem, ok := dataItems[typeAndAddr{
+			addr:   address,
+			irType: uint32(elementType),
+		}]
+		if !ok {
+			return errors.New("slice data item not found in data items")
+		}
+		var err error
+		if err = writeTokens(enc, jsontext.BeginArray); err != nil {
+			return err
+		}
+		sliceLength := int(sliceDataItem.Header().Length) / elementSize
+		sliceData := sliceDataItem.Data()
+		for i := range int(sliceLength) {
+			elementData := sliceData[i*elementSize : (i+1)*elementSize]
+			if err := d.encodeValue(enc,
+				dataItems, currentlyEncoding, v.Data.Element, elementData); err != nil {
+				return err
+			}
+		}
+		if err = writeTokens(enc, jsontext.EndArray); err != nil {
+			return err
+		}
+		return nil
+	case *ir.GoChannelType:
+		return writeTokens(enc, jsontext.String("unimplemented - channel type"))
+	case *ir.GoStringHeaderType:
+		var address uint64
+		for _, field := range v.Fields {
+			if field.Name == "str" {
+				if uint32(field.Type.GetByteSize()) != 8 || len(data) < int(field.Offset+field.Type.GetByteSize()) {
+					return errors.New("malformed string field 'str'")
+				}
+				address = binary.NativeEndian.Uint64(data[field.Offset : field.Offset+uint32(field.Type.GetByteSize())])
+			}
+		}
+		stringValue, ok := dataItems[typeAndAddr{
+			irType: uint32(v.Data.GetID()),
+			addr:   address,
+		}]
+		if !ok {
+			return fmt.Errorf("string content not present in data items")
+		}
+		length := stringValue.Header().Length
+		if len(stringValue.Data()) < int(length) {
+			return errors.New("string content not long enough for known length")
+		}
```

Comment:
If you're going to construct a new value here, give us the numbers for the lengths and the expected length.